### PR TITLE
feat: fix entity validation and typing for the metadata

### PR DESCRIPTION
**Pull request by pedrotambo**
This PR provides typing for the Entity metadata, i.e., metadata can be `Profile`, `Scene`, `Store`,  `Wearable` or `Store`. 
In addition, the ability to validate Entities using `Emote.validate` instead of manually checking the `type` field and then selecting the `<Type>.validate` function according to the `type`.
`Store` was moved to platform so all types of entities stay in the same place.
@decentraland/platform 
@decentraland/dapps

### CODE REVIEW

File: src/platform/entity.ts [hunk 1]
```diff
@@ -39,14 +47,42 @@ export namespace Entity {
...
+        else: {
+          if: { properties: { type: { const: EntityType.STORE } } },
+          then: { properties: { metadata: Store.schema } },
+          else: {
+            if: { properties: { type: { const: EntityType.EMOTE } } },
+            then: { properties: { metadata: Emote.schema } }
+          }
+        }
+      }
+    },
```
Comment (juanscolari) [inline src/platform/entity.ts:77]: Maybe something like this works here to improve readability:

```
{
  "allOf": [
    {
      "if": { ... },
      "then": { ... }
    },
    {
      "if": { ... },
      "then": { ... }
    }
  ]
}
```



Pointed out [here](https://stackoverflow.com/a/54712494/3716394)

**Review by juanscolari — APPROVED** [review]

**Review by menduz — CHANGES_REQUESTED** [review]
Can we have a helper function for this? Entity is simply a container and I think the validations of the container should be agnostic to its content. For example, adding new entities will be impossible unless we first come up with an almost final schema and commit it to master and releasing a version of almost everything using dcl-schemas.

**Comment by pedrotambo** [comment]
It's is really counterintuitive that an entity can be valid using `Entity.validate` but can have an invalid `metadata`. It breaks the logic that is preserved across the repository in all the namespaces providing `schema` and `validate`, where the `validate` validates the object and all its contents. For that reason and others i don't think that and Entity should be agnostic to its content.

If i have to add a new entity, i'd probably start here. It is not necessary to commit to master as a tag can be created.
But let's assume that's not the case and we don't want to start here. Currently, we have to set an arbitrary type of the available ones and then use whatever you want in the metadata. With this change we could achieve similar behavior using `type assertions` like `as unknown as Profile` for the metadata.  It could feel tedious but i prefer to do those things in dev code than in prod code. 
In both cases you're not going to get very far unless you mock or something the validations.

In addition, this is a general problem that we have with many repositories and i don't think that the solution would be to make things `any` or bag of random things.

Currently you need a lot of context to understand that an Entity can have anything in the metadata (from the typing view) but there's only a few set of them that would pass the validations.
There's also other cases where having the typing in the metadata would help the development, for example imagine creating a new entity object with missing fields, the compiler would prevent you from that and thus help development. 

I think that setting the `type` to string would only "help" development when creating a new entity but would do the opposite in the rest of the cases.
All of this helps would be mitigated just by creating a new type with a placeholder and publishing a development npm version. And all of the rest development cases would benefit from having a typing and the implications that this brings.
In the end, new entities development cases are far less than the rest of the cases.

Having said that, i believe you have far more context in the project and if you still think that way, i might be missing something. Could be a middleground to set the type to `string` and provide metadata validations `only` for the known types.

**Comment by pedrotambo** [comment]
I understand and agree that there would be a difference between schema validation and business logic.
The question is where is that line, and i don't understand why the entities specific schemas qualify as business logic.
The Profile/Scene/... validation functions within `common-schemas` just validate its schema. Take for example Profile, it just validates that the data contains an 'avatar' with the fields 'name', 'description', etc.
A profile that complies that schema could still be invalid, and that's where the `content-validator` kicks in: "hey, you don't own that wearable", "the hash for that file is incorrect", etc, etc.
Of course there could be some validations that might qualify as business logic. And if that's the case, they should be moved to the `content-validator` rather than saying that the whole schema is business logic.

Then if the schema should be simplified or not, or if the metadata should not be "within" the entity, that's another question.

-----
I think that if the `entity.type` could be inferred from the pointer, it could be possible to erase. But is that the case?
For emotes we said that we were going to use same urns like wearables cause they're pretty similar. Unless we add some differentiation there.

I like that ideal Entity but don't understand where would live the information that it's currently in the metadata.

That friction would `only` exist when adding new entities but no other cases (rarely).

Other dev use cases:
https://github.com/decentraland/catalyst-client/blob/master/src/utils/DeploymentBuilder.ts#L60
Here if the entity metadata has an error in the schema, the error would appear after trying to deploy to a Catalyst and all its costs.

https://github.com/decentraland/content-validator/blob/main/src/validations/ADR51.ts
We have to copy and maintain a list of the validate functions.

**Comment by menduz** [comment]
I think we have enough info. For the sake of fast forwarding I'd say that entities are agnostic containers and the validation of the metadata should live in the content-validator or specific functions

This will help us in the long run, enabling more flexibility in entities and not giving them much more power and responsibility than they should have.

In the case of emotes and wearables (collections), the rules of consensus to allow publishing are exactly the same, only the content-validator should have the responsibility of gate-keeping the publish based on the content files of the entity (including the metadata).

Derived from this, I think the content server must know nothing/not be aware of the type/schema of the metadata. Delegating the responsibility only to the content-validator

**Comment by pedrotambo** [comment]
Okay, sounds good. Do you think there's a direct action item from this or can we close this PR?

**Comment by menduz** [comment]
Maybe bring this topic to next weeks sync and close the pr for now

**Pull request closed without merging**